### PR TITLE
Lab2

### PR DESCRIPTION
**Pull request by NimishMudhale**
Fixes #42

Changes proposed in this PR: 
* Add 02_bubble/main.go for implementing BubbleSort to an integer slice
* Add 02_bubble/main_test.go which contains unit tests to verify BubbleSort implementation

### CODE REVIEW

**Review by juliaogris — COMMENTED** [review]
-   This PR should not include any files from `00_fib/`
-   Please fix up PR description according to [template](https://github.com/anz-bank/go-training/blob/master/.github/pull_request_template.md)
-   Please edit your commits to follow [recommendations](https://chris.beams.io/posts/git-commit/)
    -   Capitalize the subject line
    -   Use the imperative mood in the subject line
    -   Fold fixup commits into previous ones with `git rebase -i HASH`

**Comment by mohankrishna** [comment]
Please check your folder structure. The folder structure should be `02_bubble/NimishMudhale` please create a new folder inside 02_bubble with name `NimishMudhale`and move the code files inside that. This will help us keep all the lab2 related code inside one folder.

File: 02_Lab2_NimishMudhale/bubble_sort.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	sorted := false
+	for !sorted {
+		swapped := false
+		for i := 0; i < n-1; i++ {
+			if s[i] > s[i+1] {
+				s[i+1], s[i] = s[i], s[i+1]
+				swapped = true
+			}
+		}
+		if !swapped {
```
Comment (mohankrishna) [inline 02_Lab2_NimishMudhale/bubble_sort.go]: You can directly use `swapped` in for loop. When there are no swaps that means the array is sorted.

File: 02_Lab2_NimishMudhale/bubble_sort.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	n := len(s)
+	sorted := false
+	for !sorted {
+		swapped := false
```
Comment (anzdaddy) [inline 02_Lab2_NimishMudhale/bubble_sort.go]: It looks like `swapped` and `sorted` are basically opposites of each other. Remove one.

File: 0_Fib/fib.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package main
```
Comment (anzdaddy) [inline 0_Fib/fib.go]: Why is fib.go in this PR?

File: 0_Fib/fib_test.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package main
```
Comment (anzdaddy) [inline 0_Fib/fib_test.go]: Why is fib_test.go in this PR?

**Review by anzdaddy — CHANGES_REQUESTED** [review]

File: 02_Lab2_NimishMudhale/bubble_sort.go (hunk 1 above)
Comment (NimishMudhale) [reply 02_Lab2_NimishMudhale/bubble_sort.go]: done

File: 02_Lab2_NimishMudhale/bubble_sort.go (hunk 2 above)
Comment (NimishMudhale) [reply 02_Lab2_NimishMudhale/bubble_sort.go]: Done

File: 02_bubble/NimishMudhale/bubble_sort.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+				s[i+1], s[i] = s[i], s[i+1]
+				swapped = true
+			}
+		}
+		if !swapped {
+			sorted = true
+		}
+		n--
+	}
+	return s
```
Comment (anzdaddy) [inline 02_bubble/NimishMudhale/bubble_sort.go]: Can't mutate the original. Make a copy. But first write a test that detects the lack of copying.

File: 02_bubble/NimishMudhale/bubble_sort.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+}
+func main() {
```
Comment (anzdaddy) [inline 02_bubble/NimishMudhale/bubble_sort.go:29]: Put a blank line between functions.

File: 02_bubble/NimishMudhale/bubble_sort_test.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+}
+func TestBubbleSortNegativeValues(t *testing.T) {
+	//Given
+	assert := assert.New(t)
+	//when
+	sortedArray := bubble([]int{-3, -2, -1, -5})
+
+	//Then
+	assert.Equal([]int{-5, -3, -2, -1}, sortedArray)
+
```
Comment (anzdaddy) [inline 02_bubble/NimishMudhale/bubble_sort_test.go:32]: Unnecessary blank line

**Review by anzdaddy — CHANGES_REQUESTED** [review]

**Comment by juliaogris** [comment]
Click `Update branch` and locally `git pull origin lab2` before adding new commits.

File: 02_bubble/NimishMudhale/bubble_sort.go (hunk 5 above)
Comment (NimishMudhale) [reply 02_bubble/NimishMudhale/bubble_sort.go]: done

File: 02_bubble/NimishMudhale/bubble_sort.go (hunk 6 above)
Comment (NimishMudhale) [reply 02_bubble/NimishMudhale/bubble_sort.go:29]: done

File: 02_bubble/NimishMudhale/bubble_sort_test.go (hunk 7 above)
Comment (NimishMudhale) [reply 02_bubble/NimishMudhale/bubble_sort_test.go:32]: done

File: 02_bubble/NimishMudhale/bubble_sort.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	sorted := make([]int, len(s))
+	copy(sorted, s)
+	n := len(s)
+	for {
+		if n == 0 {
```
Comment (juliaogris) [inline 02_bubble/NimishMudhale/bubble_sort.go:16]: I'd rework this to (mostly a matter of taste):
```go
func bubble(s []int) []int {
	sorted := make([]int, len(s))
	copy(sorted, s)
	n := len(s)
        if (n < 2) {
              return sorted
       }
       for { //...
```

File: 02_bubble/NimishMudhale/bubble_sort_test.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	expected := "[1 2 3 5]"
+	actual := buf.String()
+	r.Equalf(expected, actual, "Unexpected output in main()")
+}
+func TestBubbleSortNegativeValues(t *testing.T) {
+	//Given
+	assert := assert.New(t)
+	//when
+	input := []int{-3, -2, -1, -5}
+	sortedArray := bubble(input)
```
Comment (juliaogris) [inline 02_bubble/NimishMudhale/bubble_sort_test.go:28]: Nice. Use a few more input/output/expected datasets with [table driven tests](https://github.com/golang/go/wiki/TableDrivenTests), especially corner cases: empty slice, one element slice, slice with duplicated element.
Also ensure that a real copy is made of input.

Also, please rework your commit messages - there should be no mention of Fibonacci.
1-3 commit messages is probably sufficient for this PR.
-   Capitalize the subject line
-   Use the imperative mood in the subject line
-   Fold fixup commits into previous ones with `git rebase -i HASH`
    See [slides](https://gotraining.xyz/gotraining.slide#29).

**Review by juliaogris — CHANGES_REQUESTED** [review]

**Comment by juliaogris** [comment]
Hi,
There has been  no activity in the last 20 days so I will close this PR.
Feel free to open it again if you have something to add.
Thanks!
Julia

**Pull request closed without merging**